c++/src/kj/compat/tls-test.c++

```diff
@@ -818,6 +822,241 @@ KJ_TEST("TLS client certificate verification") {
+};
+
+KJ_TEST("TLS receiver basics") {
+  TlsReceiverTest test;
+  ErrorNexus e;
+
+  test.receiver->run();
+
+  auto clientPromise = test.baseReceiver->connect()
+                           .then([&](auto stream) {
+                             return test.tlsClient.wrapClient(kj::mv(stream), "example.com");
+                           })
+                           .then([&](auto client) {
+                             return test.writeToServer(kj::mv(client));
+                           });
+
+  auto serverPromise = test.receiver->accept().then([&](auto server) {
+    return test.readFromClient(kj::mv(server));
+  });
+
+  clientPromise.wait(test.io.waitScope);
+  serverPromise.wait(test.io.waitScope);
+}
+
+KJ_TEST("TLS receiver experiences pre-TLS error") {
+  TlsReceiverTest test;
+  ErrorNexus e;
+
+  test.receiver->run();
+
+  KJ_LOG(INFO, "Disappointing our server");
+  KJ_EXPECT_LOG(ERROR, "TlsConnectionReceiver could not accept()");
+  test.baseReceiver->badConnect();
+
+  KJ_LOG(INFO, "Trying again with a good connection");
+  auto clientPromise = test.baseReceiver->connect()
+                           .then([&](auto stream) {
+                             return test.tlsClient.wrapClient(kj::mv(stream), "example.com");
+                           })
+                           .then([&](auto client) {
+                             return test.writeToServer(kj::mv(client));
+                           });
+
+  auto serverPromise = test.receiver->accept().then([&](auto server) {
+    return test.readFromClient(kj::mv(server));
+  });
+
+  clientPromise.wait(test.io.waitScope);
+  serverPromise.wait(test.io.waitScope);
+}
+
+KJ_TEST("TLS receiver accepts multiple clients") {
+  TlsReceiverTest test;
+  ErrorNexus e;
+
+  test.receiver->run();
+
+  KJ_LOG(INFO, "Requesting a bunch of client connects");
+  constexpr auto kClientCount = 20;
+  auto clientPromises = Vector<Promise<void>>();
+  for (auto i = 0; i < kClientCount; ++i) {
+    auto promise = test.baseReceiver->connect()
+                       .then([&](auto stream) {
+                         return test.tlsClient.wrapClient(kj::mv(stream), "example.com");
+                       })
+                       .then([&](auto client) {
+                         return test.writeToServer(kj::mv(client));
+                       });
+    clientPromises.add(kj::mv(promise));
+  }
+
+  KJ_LOG(INFO, "Requesting and resolving a bunch of server accepts in sequence");
+  for (auto i = 0; i < kClientCount; ++i) {
+    // Resolve each receive in sequence like the Supervisor/Network.
+    test.receiver->accept()
+        .then([&](auto server) {
+          return test.readFromClient(kj::mv(server));
+        })
+        .wait(test.io.waitScope);
+  }
+
+  KJ_LOG(INFO, "Resolving all of our client connects in parallel");
+  joinPromises(clientPromises.releaseAsArray()).wait(test.io.waitScope);
+
+  KJ_LOG(INFO, "Requesting one last server accept that we'll never resolve");
```

Comment:
If we want to at least minimally test that this doesn't resolve, you could add a `KJ_EXPECT(!extraAcceptPromise.poll(test.io.waitScope));` to the last line of the test.

Reply:
That sounds like an improvement to me!